benchmarks/random_allocations/random_allocations.cpp

```diff
@@ -0,0 +1,181 @@
+  std::uniform_int_distribution<int> index_distribution(0, num_allocations-1);
+
+  int active_allocations{0};
+  int allocation_count{0};
+
+  allocation_vector allocations{};
+  size_t allocation_size{0};
+
+  for (int i = 0; i < num_allocations * 2; ++i) {
+    bool do_alloc = true;
+    size_t size = static_cast<size_t>(size_distribution(generator));
+    
+    if (active_allocations > 0) {
+      int chance = op_distribution(generator);
+      do_alloc = (chance < allocation_probability) &&
+                 (allocation_count < num_allocations) &&
+                 (allocation_size + size < max_usage);
+    }
+
+    void* ptr = nullptr;
+    if (do_alloc) { // try to allocate
+      try {
+        ptr = mr.allocate(size, stream);
+      } catch(std::bad_alloc) {
+        do_alloc = false;
+      }
+    }
+
+    if (do_alloc) { // alloc succeeded
+      allocations.emplace_back(ptr, size);
+      active_allocations++;
+      allocation_count++;
+      allocation_size += size;
+
+      #if VERBOSE
+      std::cout << active_allocations << " | " << allocation_count << " Allocating: " << size 
+                << " | total: " << allocation_size << "\n";
+      #endif
+    }
+    else { // dealloc, or alloc failed
+      if (active_allocations > 0) {
+        size_t index = index_distribution(generator) % active_allocations;
+        active_allocations--;
+        allocation to_free = remove_at(allocations, index);
+        mr.deallocate(to_free.p, to_free.size, stream);
+        allocation_size -= to_free.size;
+
+        #if VERBOSE
+        std::cout << active_allocations << " | " << allocation_count << " Deallocating: " 
+                  << to_free.size << " at " << index << " | total: " << allocation_size << "\n";
+        #endif
+      }
+    }
+  }
+
+  assert(active_allocations == 0);
+  assert(allocations.size() == 0);
+}
+}  // namespace
+
+void uniform_random_allocations(rmm::mr::device_memory_resource& mr,
+                                size_t num_allocations,
+                                size_t max_allocation_size, // in MiB
+                                size_t max_usage,
+                                cudaStream_t stream = 0) {
+  std::uniform_int_distribution<std::size_t> size_distribution(1, max_allocation_size * size_mb);
+  random_allocation_free(mr, size_distribution, num_allocations, max_usage, stream);
+}
+
+// TODO figure out how to map a normal distribution to integers between 1 and max_allocation_size
+/*void normal_random_allocations(rmm::mr::device_memory_resource& mr,
+                                size_t num_allocations = 1000,
+                                size_t mean_allocation_size = 500, // in MiB
+                                size_t stddev_allocation_size = 500, // in MiB
+                                size_t max_usage = 8 << 20,
+                                cudaStream_t stream) {
+  std::normal_distribution<std::size_t> size_distribution(, max_allocation_size * size_mb);
```

Comment:
Couldn't we just cast the `double` to an integer? Technically that would make the distribution not a statistically perfect "normal" distribution, but I think that's okay for our purposes here.

Reply:
And clamp to 0, max_allocation_size?

Reply:
I think for our purposes we may not even need this test.